File: nodejs-backend/src/models/Membership.js
/**
 * Membership Model
 * Converted from Spring Boot JPA Entity
 */

const { DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  const Membership = sequelize.define('Membership', {
    id: {
      type: DataTypes.BIGINT,
      primaryKey: true,
      autoIncrement: true
    },
    name: {
      type: DataTypes.STRING(255),
      allowNull: false,
      validate: {
        notEmpty: true
      }
    },
    description: {
      type: DataTypes.TEXT
    },
    price: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false
    },
    duration_days: {
      type: DataTypes.INTEGER,
      defaultValue: 30,
      field: 'duration_days'
    },
    benefits: {
      type: DataTypes.TEXT,
      get() {
        const rawValue = this.getDataValue('benefits');
        if (!rawValue) return null;
        // Check if it's JSON or comma-separated
        try {
          return JSON.parse(rawValue);
        } catch {
          return rawValue.split(',').map(b => b.trim());
        }
      },
      set(value) {
        if (Array.isArray(value)) {
          this.setDataValue('benefits', value.join(','));
        } else {
          this.setDataValue('benefits', value);
        }
      }
    },
    is_active: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
      field: 'is_active'
    },
    created_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW,
      field: 'created_at'
    },
    updated_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW,
      field: 'updated_at'
    }
  }, {
    tableName: 'memberships',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  return Membership;
};

File: flutter_app/lib/models/membership.dart
/// Membership model for KJX Esports app
class Membership {
  final int? id;
  final String name;
  final String? description;
  final double price;
  final int durationMonths;
  final String? benefits;
  final bool isActive;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Membership({
    this.id,
    required this.name,
    this.description,
    required this.price,
    this.durationMonths = 1,
    this.benefits,
    this.isActive = true,
    this.createdAt,
    this.updatedAt,
  });

  factory Membership.fromJson(Map<String, dynamic> json) {
    return Membership(
      id: json['id'],
      name: json['name'] ?? '',
      description: json['description'],
      price: (json['price'] is String)
          ? double.parse(json['price'])
          : (json['price'] ?? 0).toDouble(),
      durationMonths: json['duration_months'] ?? json['durationMonths'] ?? 1,
      benefits: json['benefits'],
      isActive: json['is_active'] ?? json['isActive'] ?? true,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'name': name,
      if (description != null) 'description': description,
      'price': price,
      'duration_months': durationMonths,
      if (benefits != null) 'benefits': benefits,
      'is_active': isActive,
    };
  }

  List<String> get benefitsList {
    if (benefits == null || benefits!.isEmpty) return [];
    return benefits!.split(', ');
  }

  bool get isLifetime => durationMonths == 0;

  Membership copyWith({
    int? id,
    String? name,
    String? description,
    double? price,
    int? durationMonths,
    String? benefits,
    bool? isActive,
  }) {
    return Membership(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      price: price ?? this.price,
      durationMonths: durationMonths ?? this.durationMonths,
      benefits: benefits ?? this.benefits,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }

  @override
  String toString() => 'Membership(id: $id, name: $name, price: $price)';
}
